reframe/core/logging.py

```diff
@@ -83,50 +87,163 @@ def set_handler_level(hdlr, level):
+        return logging.StreamHandler(stream=sys.stderr)
+    else:
+        raise ConfigError('unknown stream: %s' % stream)
+
+
+def _create_graylog_handler(handler_config):
+    try:
+        import pygelf
+    except ImportError:
+        return None
+
+    hostname = handler_config.get('hostname', None)
+    port = handler_config.get('port', None)
+    facility = handler_config.get('facility', None)
+    return pygelf.GelfHttpHandler(host=hostname,
+                                  port=port,
+                                  facility=facility,
+                                  debug=True,
+                                  include_extra_fields=True)
+
+
+def _extract_handlers(handlers_list, auto_logfile=None):
+    # Check if we are using the old syntax
+    if isinstance(handlers_list, collections.abc.Mapping):
+        handlers_list = _convert_handler_syntax(handlers_list)
+        sys.stderr.write(
+            'WARNING: looks like you are using an old syntax for '
+            'logging configuration; please update your syntax as follows:\n'
+            '\nhandlers: %s\n' % pprint.pformat(handlers_list, indent=1))
+
     handlers = []
-    if not handlers_dict:
-        raise ValueError('no handlers are defined for logger')
+    if not handlers_list:
+        raise ValueError('no handlers are defined for logging')
 
-    for filename, handler_config in handlers_dict.items():
+    for i, handler_config in enumerate(handlers_list):
         if not isinstance(handler_config, collections.abc.Mapping):
-            raise TypeError('handler %s is not a dictionary' % filename)
+            raise TypeError('handler config at position %s '
+                            'is not a dictionary' % i)
 
-        level = handler_config.get('level', 'debug').lower()
-        fmt   = handler_config.get('format', '%(message)s')
-        datefmt = handler_config.get('datefmt', '%FT%T')
-        append  = handler_config.get('append', False)
-        timestamp = handler_config.get('timestamp', None)
-
-        if filename == '&1':
-            hdlr = logging.StreamHandler(stream=sys.stdout)
-        elif filename == '&2':
-            hdlr = logging.StreamHandler(stream=sys.stderr)
+        try:
+            handler_type = handler_config['type']
+        except KeyError:
+            raise ConfigError('no type specified for '
+                              'handler at position %s' % i) from None
+
+        if handler_type == 'file':
+            hdlr = _create_file_handler(handler_config)
+        elif handler_type == 'dynfile':
+            hdlr = _create_dynfile_handler(handler_config)
+        elif handler_type == 'stream':
+            hdlr = _create_stream_handler(handler_config)
+        elif handler_type == 'graylog':
+            hdlr = _create_graylog_handler(handler_config)
+            if hdlr is None:
+                sys.stderr.write('WARNING: could not initialize the '
+                                 'graylog handler; ignoring...\n')
```

Comment:
I am getting this warning, which means that `Python-bare` packages does not contain it. We should add it there.